File: Python/Search-Insert-Position.py
#!/usr/bin/env python

class Solution:
    # @param A, a list of integers
    # @param target, an integer to be inserted
    # @return integer
    def searchInsert(self, A, target):
        left = 0
        right = len(A) - 1
        while left <= right:
            mid = (left + right)/2
            if A[mid] == target:
                return mid
            elif A[mid] < target:
                if mid == right:
                    return mid + 1
                elif A[mid+1] > target:
                    return mid+1
                else:
                    left = mid + 1
            elif A[mid] > target:
                if mid == 0:
                    return 0
                elif A[mid-1] < target:
                    return mid
                else:
                    right = mid - 1;

File: C++/Search-Insert-Position.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
    	if (nums.empty()) return 0;
        int left = 0, right = nums.size() - 1;
        while (left + 1 < right) {
        	int mid = left + (right - left) / 2;
        	if (nums[mid] >= target) right = mid;
        	else left = mid;
        }
        if (nums[left] >= target) return left;
        else if (nums[right] >= target) return right;
        return nums.size();
    }
};